File: utils/src/main/java/com/bharathksunil/utils/TextDrawable.java
package com.bharathksunil.utils;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.OvalShape;
import android.graphics.drawable.shapes.RectShape;
import android.graphics.drawable.shapes.RoundRectShape;

/**
 * TextDrawable This light-weight library provides images with letter/text like the Gmail app
 * This class is imported from com.amulyakhare:com.amulyakhare.textdrawable:1.0.1
 *
 * @author Bharath on 15-08-2017.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public final class TextDrawable extends ShapeDrawable {

    private static final float SHADE_FACTOR = 0.9f;
    private final Paint mTextPaint;
    private final Paint mBorderPaint;
    private final String mText;
    @SuppressWarnings("FieldCanBeLocal")
    private final int mColor;
    private final RectShape mShape;
    private final int mDrawableHeight;
    private final int mDrawableWidth;
    private final int mTextFontSize;
    private final float mDrawableRadius;
    private final int mDrawableBorderThickness;

    //todo: Add documentation and code cleanup
    private TextDrawable(Builder builder) {
        super(builder.shape);

        // shape properties
        mShape = builder.shape;
        mDrawableHeight = builder.height;
        mDrawableWidth = builder.width;
        mDrawableRadius = builder.mDrawableRadius;

        // text and color
        mText = builder.toUpperCase ? builder.mDrawableText.toUpperCase() : builder.mDrawableText;
        mColor = builder.color;

        // text paint settings
        mTextFontSize = builder.fontSize;
        mTextPaint = new Paint();
        mTextPaint.setColor(builder.mTextColor);
        mTextPaint.setAntiAlias(true);
        mTextPaint.setFakeBoldText(builder.isBold);
        mTextPaint.setStyle(Paint.Style.FILL);
        mTextPaint.setTypeface(builder.font);
        mTextPaint.setTextAlign(Paint.Align.CENTER);
        mTextPaint.setStrokeWidth(builder.borderThickness);

        // border paint settings
        mDrawableBorderThickness = builder.borderThickness;
        mBorderPaint = new Paint();
        mBorderPaint.setColor(getDarkerShade(mColor));
        mBorderPaint.setStyle(Paint.Style.STROKE);
        mBorderPaint.setStrokeWidth(mDrawableBorderThickness);

    }

    public static IShapeBuilder builder() {
        return new Builder();
    }

    private int getDarkerShade(int color) {
        return Color.rgb((int) (SHADE_FACTOR * Color.red(color)),
                (int) (SHADE_FACTOR * Color.green(color)),
                (int) (SHADE_FACTOR * Color.blue(color)));
    }

    @Override
    public void draw(Canvas canvas) {
        super.draw(canvas);
        Rect r = getBounds();


        // draw border
        if (mDrawableBorderThickness > 0) {
            drawBorder(canvas);
        }

        int count = canvas.save();
        canvas.translate(r.left, r.top);

        // draw text
        int width = this.mDrawableWidth < 0 ? r.width() : this.mDrawableWidth;
        int height = this.mDrawableHeight < 0 ? r.height() : this.mDrawableHeight;
        int fontSize = this.mTextFontSize < 0 ? (Math.min(width, height) / 2) : this.mTextFontSize;
        mTextPaint.setTextSize(fontSize);
        canvas.drawText(mText, width / 2, height / 2 - ((mTextPaint.descent() + mTextPaint.ascent()) / 2), mTextPaint);

        canvas.restoreToCount(count);

    }

    private void drawBorder(Canvas canvas) {
        RectF rect = new RectF(getBounds());
        rect.inset(mDrawableBorderThickness / 2, mDrawableBorderThickness / 2);

        if (mShape instanceof OvalShape) {
            canvas.drawOval(rect, mBorderPaint);
        } else if (mShape instanceof RoundRectShape) {
            canvas.drawRoundRect(rect, mDrawableRadius, mDrawableRadius, mBorderPaint);
        } else {
            canvas.drawRect(rect, mBorderPaint);
        }
    }

    @Override
    public void setAlpha(int alpha) {
        mTextPaint.setAlpha(alpha);
    }

    @Override
    public void setColorFilter(ColorFilter cf) {
        mTextPaint.setColorFilter(cf);
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSLUCENT;
    }

    @Override
    public int getIntrinsicWidth() {
        return mDrawableWidth;
    }

    @Override
    public int getIntrinsicHeight() {
        return mDrawableHeight;
    }

    public interface IConfigBuilder {
        IConfigBuilder width(int width);

        IConfigBuilder height(int height);

        IConfigBuilder textColor(int color);

        IConfigBuilder withBorder(int thickness);

        IConfigBuilder useFont(Typeface font);

        IConfigBuilder fontSize(int size);

        IConfigBuilder bold();

        IConfigBuilder toUpperCase();

        IShapeBuilder endConfig();
    }

    public interface IBuilder {

        TextDrawable build(String text, int color);
    }

    @SuppressWarnings("UnusedReturnValue")
    public interface IShapeBuilder {

        IConfigBuilder beginConfig();

        IBuilder rect();

        IBuilder round();

        IBuilder roundRect(int radius);

        TextDrawable buildRect(String text, int color);

        TextDrawable buildRoundRect(String text, int color, int radius);

        TextDrawable buildRound(String text, int color);
    }

    private static class Builder implements IConfigBuilder, IShapeBuilder, IBuilder {

        private int mTextColor;
        float mDrawableRadius;
        private String mDrawableText;
        private int color;
        private int borderThickness;
        private int width;
        private int height;
        private Typeface font;
        private RectShape shape;
        private int fontSize;
        private boolean isBold;
        private boolean toUpperCase;

        private Builder() {
            mDrawableText = "";
            color = Color.GRAY;
            mTextColor = Color.WHITE;
            borderThickness = 0;
            width = -1;
            height = -1;
            shape = new RectShape();
            font = Typeface.create("sans-serif-light", Typeface.NORMAL);
            fontSize = -1;
            isBold = false;
            toUpperCase = false;
        }

        public IConfigBuilder width(int width) {
            this.width = width;
            return this;
        }

        public IConfigBuilder height(int height) {
            this.height = height;
            return this;
        }

        public IConfigBuilder textColor(int color) {
            this.mTextColor = color;
            return this;
        }

        public IConfigBuilder withBorder(int thickness) {
            this.borderThickness = thickness;
            return this;
        }

        public IConfigBuilder useFont(Typeface font) {
            this.font = font;
            return this;
        }

        public IConfigBuilder fontSize(int size) {
            this.fontSize = size;
            return this;
        }

        public IConfigBuilder bold() {
            this.isBold = true;
            return this;
        }

        public IConfigBuilder toUpperCase() {
            this.toUpperCase = true;
            return this;
        }

        @Override
        public IConfigBuilder beginConfig() {
            return this;
        }

        @Override
        public IShapeBuilder endConfig() {
            return this;
        }

        @Override
        public IBuilder rect() {
            this.shape = new RectShape();
            return this;
        }

        @Override
        public IBuilder round() {
            this.shape = new OvalShape();
            return this;
        }

        @Override
        public IBuilder roundRect(int radius) {
            this.mDrawableRadius = radius;
            float[] radii = {radius, radius, radius, radius, radius, radius, radius, radius};
            this.shape = new RoundRectShape(radii, null, null);
            return this;
        }

        @Override
        public TextDrawable buildRect(String text, int color) {
            rect();
            return build(text, color);
        }

        @Override
        public TextDrawable buildRoundRect(String text, int color, int radius) {
            roundRect(radius);
            return build(text, color);
        }

        @Override
        public TextDrawable buildRound(String text, int color) {
            round();
            return build(text, color);
        }

        @Override
        public TextDrawable build(String text, int color) {
            this.color = color;
            this.mDrawableText = text;
            return new TextDrawable(this);
        }
    }
}

File: utils/src/main/java/com/bharathksunil/utils/TextDrawable.kt
package com.bharathksunil.utils

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ColorFilter
import android.graphics.Paint
import android.graphics.PixelFormat
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.Typeface
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.OvalShape
import android.graphics.drawable.shapes.RectShape
import android.graphics.drawable.shapes.RoundRectShape

/**
 * TextDrawable This light-weight library provides images with letter/text like the Gmail app
 * This class is imported from com.amulyakhare:com.amulyakhare.textdrawable:1.0.1
 *
 * @author Bharath on 15-08-2017.
 */
class TextDrawable//todo: Add documentation and code cleanup
private constructor(builder: Builder) : ShapeDrawable(builder.shape) {
    private val mTextPaint: Paint
    private val mBorderPaint: Paint
    private val mText: String?
    private val mColor: Int
    private val mShape: RectShape?
    private val mDrawableHeight: Int
    private val mDrawableWidth: Int
    private val mTextFontSize: Int
    private val mDrawableRadius: Float
    private val mDrawableBorderThickness: Int

    init {

        // shape properties
        mShape = builder.shape
        mDrawableHeight = builder.height
        mDrawableWidth = builder.width
        mDrawableRadius = builder.mDrawableRadius

        // text and color
        mText = if (builder.toUpperCase) builder.mDrawableText!!.toUpperCase() else builder.mDrawableText
        mColor = builder.color

        // text paint settings
        mTextFontSize = builder.fontSize
        mTextPaint = Paint()
        mTextPaint.color = builder.mTextColor
        mTextPaint.isAntiAlias = true
        mTextPaint.isFakeBoldText = builder.isBold
        mTextPaint.style = Paint.Style.FILL
        mTextPaint.typeface = builder.font
        mTextPaint.textAlign = Paint.Align.CENTER
        mTextPaint.strokeWidth = builder.borderThickness.toFloat()

        // border paint settings
        mDrawableBorderThickness = builder.borderThickness
        mBorderPaint = Paint()
        mBorderPaint.color = getDarkerShade(mColor)
        mBorderPaint.style = Paint.Style.STROKE
        mBorderPaint.strokeWidth = mDrawableBorderThickness.toFloat()

    }

    private fun getDarkerShade(color: Int): Int {
        return Color.rgb((SHADE_FACTOR * Color.red(color)).toInt(),
                (SHADE_FACTOR * Color.green(color)).toInt(),
                (SHADE_FACTOR * Color.blue(color)).toInt())
    }

    override fun draw(canvas: Canvas) {
        super.draw(canvas)
        val r = bounds


        // draw border
        if (mDrawableBorderThickness > 0) {
            drawBorder(canvas)
        }

        val count = canvas.save()
        canvas.translate(r.left.toFloat(), r.top.toFloat())

        // draw text
        val width = if (this.mDrawableWidth < 0) r.width() else this.mDrawableWidth
        val height = if (this.mDrawableHeight < 0) r.height() else this.mDrawableHeight
        val fontSize = if (this.mTextFontSize < 0) Math.min(width, height) / 2 else this.mTextFontSize
        mTextPaint.textSize = fontSize.toFloat()
        canvas.drawText(mText!!, (width / 2).toFloat(), height / 2 - (mTextPaint.descent() + mTextPaint.ascent()) / 2, mTextPaint)

        canvas.restoreToCount(count)

    }

    private fun drawBorder(canvas: Canvas) {
        val rect = RectF(bounds)
        rect.inset((mDrawableBorderThickness / 2).toFloat(), (mDrawableBorderThickness / 2).toFloat())

        if (mShape is OvalShape) {
            canvas.drawOval(rect, mBorderPaint)
        } else if (mShape is RoundRectShape) {
            canvas.drawRoundRect(rect, mDrawableRadius, mDrawableRadius, mBorderPaint)
        } else {
            canvas.drawRect(rect, mBorderPaint)
        }
    }

    override fun setAlpha(alpha: Int) {
        mTextPaint.alpha = alpha
    }

    override fun setColorFilter(cf: ColorFilter?) {
        mTextPaint.colorFilter = cf
    }

    override fun getOpacity(): Int {
        return PixelFormat.TRANSLUCENT
    }

    override fun getIntrinsicWidth(): Int {
        return mDrawableWidth
    }

    override fun getIntrinsicHeight(): Int {
        return mDrawableHeight
    }

    interface IConfigBuilder {
        fun width(width: Int): IConfigBuilder

        fun height(height: Int): IConfigBuilder

        fun textColor(color: Int): IConfigBuilder

        fun withBorder(thickness: Int): IConfigBuilder

        fun useFont(font: Typeface): IConfigBuilder

        fun fontSize(size: Int): IConfigBuilder

        fun bold(): IConfigBuilder

        fun toUpperCase(): IConfigBuilder

        fun endConfig(): IShapeBuilder
    }

    interface IBuilder {

        fun build(text: String, color: Int): TextDrawable
    }

    interface IShapeBuilder {

        fun beginConfig(): IConfigBuilder

        fun rect(): IBuilder

        fun round(): IBuilder

        fun roundRect(radius: Int): IBuilder

        fun buildRect(text: String, color: Int): TextDrawable

        fun buildRoundRect(text: String, color: Int, radius: Int): TextDrawable

        fun buildRound(text: String, color: Int): TextDrawable
    }

    private class Builder private constructor() : IConfigBuilder, IShapeBuilder, IBuilder {

        private var mTextColor: Int = 0
        internal var mDrawableRadius: Float = 0.toFloat()
        private var mDrawableText: String? = null
        private var color: Int = 0
        private var borderThickness: Int = 0
        private var width: Int = 0
        private var height: Int = 0
        private var font: Typeface? = null
        private var shape: RectShape? = null
        private var fontSize: Int = 0
        private var isBold: Boolean = false
        private var toUpperCase: Boolean = false

        init {
            mDrawableText = ""
            color = Color.GRAY
            mTextColor = Color.WHITE
            borderThickness = 0
            width = -1
            height = -1
            shape = RectShape()
            font = Typeface.create("sans-serif-light", Typeface.NORMAL)
            fontSize = -1
            isBold = false
            toUpperCase = false
        }

        override fun width(width: Int): IConfigBuilder {
            this.width = width
            return this
        }

        override fun height(height: Int): IConfigBuilder {
            this.height = height
            return this
        }

        override fun textColor(color: Int): IConfigBuilder {
            this.mTextColor = color
            return this
        }

        override fun withBorder(thickness: Int): IConfigBuilder {
            this.borderThickness = thickness
            return this
        }

        override fun useFont(font: Typeface): IConfigBuilder {
            this.font = font
            return this
        }

        override fun fontSize(size: Int): IConfigBuilder {
            this.fontSize = size
            return this
        }

        override fun bold(): IConfigBuilder {
            this.isBold = true
            return this
        }

        override fun toUpperCase(): IConfigBuilder {
            this.toUpperCase = true
            return this
        }

        override fun beginConfig(): IConfigBuilder {
            return this
        }

        override fun endConfig(): IShapeBuilder {
            return this
        }

        override fun rect(): IBuilder {
            this.shape = RectShape()
            return this
        }

        override fun round(): IBuilder {
            this.shape = OvalShape()
            return this
        }

        override fun roundRect(radius: Int): IBuilder {
            this.mDrawableRadius = radius.toFloat()
            val radii = floatArrayOf(radius.toFloat(), radius.toFloat(), radius.toFloat(), radius.toFloat(), radius.toFloat(), radius.toFloat(), radius.toFloat(), radius.toFloat())
            this.shape = RoundRectShape(radii, null, null)
            return this
        }

        override fun buildRect(text: String, color: Int): TextDrawable {
            rect()
            return build(text, color)
        }

        override fun buildRoundRect(text: String, color: Int, radius: Int): TextDrawable {
            roundRect(radius)
            return build(text, color)
        }

        override fun buildRound(text: String, color: Int): TextDrawable {
            round()
            return build(text, color)
        }

        override fun build(text: String, color: Int): TextDrawable {
            this.color = color
            this.mDrawableText = text
            return TextDrawable(this)
        }
    }

    companion object {

        private val SHADE_FACTOR = 0.9f

        fun builder(): IShapeBuilder {
            return Builder()
        }
    }
}
